File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/AdminSpoofUserGameCommand.vb
Imports AtlasV.Daemon
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class AdminSpoofUserGameCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim t = If(Arguments.Count = 0, "", Arguments(0))
            Dim r As String
            Dim target As GameState = Nothing

            If t.Length = 0 OrElse Not Battlenet.Common.GetClientByOnlineName(t, target) OrElse target Is Nothing Then
                r = Resources.UserNotLoggedOn

                For Each line In r.Split(Battlenet.Common.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            Arguments.RemoveAt(0)
            'RawBuffer = RawBuffer((Encoding.UTF8.GetByteCount(t) + (If(Arguments.Count > 0, 1, 0)))..)
            RawBuffer = RawBuffer.Skip(Encoding.UTF8.GetByteCount(t) + (If(Arguments.Count > 0, 1, 0))).ToArray()
            Dim strGame = If(Arguments.Count = 0, "", Arguments(0))
            Dim targetGame = Product.StringToProduct(strGame)

            If targetGame = Product.ProductCode.None Then
                r = Resources.AdminSpoofUserGameCommandBadValue

                For Each line In r.Split(Battlenet.Common.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            SyncLock target.Statstring
                Dim newStatstring = New Byte(target.Statstring.Length - 1) {}

                Using m = New MemoryStream(newStatstring)
                    Using w = New BinaryWriter(m)
                        w.Write(CUInt(targetGame))
                        w.Write(target.Statstring.Skip(4).ToArray()) 'target.Statstring(4..))
                    End Using
                End Using

                If target.ActiveChannel IsNot Nothing Then
                    target.ActiveChannel.UpdateUser(target, newStatstring)
                Else
                    target.Statstring = newStatstring
                End If
            End SyncLock

            Dim targetEnv = New Dictionary(Of String, String)() From {
                {"accountName", target.Username},
                {"channel", If(target.ActiveChannel Is Nothing, "(null)", target.ActiveChannel.Name)},
                {"game", Product.ProductName(targetGame, True)},
                {"host", "BNETDocs"},
                {"localTime", target.LocalTime.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ")},
                {"name", target.OnlineName},
                {"onlineName", target.OnlineName},
                {"realm", "BNETDocs"},
                {"realmTime", DateTime.Now.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ")},
                {"realmTimezone", $"UTC{DateTime.Now}"},
                {"user", target.OnlineName},
                {"username", target.OnlineName},
                {"userName", target.OnlineName}
            }
            Dim env = targetEnv.Concat(varContext.Environment)
            r = Resources.AdminSpoofUserGameCommand

            For Each kv In env
                r = r.Replace("{" & kv.Key & "}", kv.Value)
            Next

            For Each line In r.Split(Battlenet.Common.NewLine)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
            Next
        End Sub
    End Class
End Namespace

File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/AdminSpoofUserGameCommand.cs
using Atlasd.Daemon;
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class AdminSpoofUserGameCommand : ChatCommand
    {
        public AdminSpoofUserGameCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            var t = Arguments.Count == 0 ? "" : Arguments[0]; // target
            string r; // reply

            if (t.Length == 0 || !Battlenet.Common.GetClientByOnlineName(t, out var target) || target == null)
            {
                r = Resources.UserNotLoggedOn;
                foreach (var line in r.Split(Battlenet.Common.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            Arguments.RemoveAt(0); // remove target
            // Calculates and removes (target+' ') from (raw) which prints into (newRaw):
            RawBuffer = RawBuffer[(Encoding.UTF8.GetByteCount(t) + (Arguments.Count > 0 ? 1 : 0))..];

            var strGame = Arguments.Count == 0 ? "" : Arguments[0];
            var targetGame = Product.StringToProduct(strGame);
            if (targetGame == Product.ProductCode.None)
            {
                r = Resources.AdminSpoofUserGameCommandBadValue;
                foreach (var line in r.Split(Battlenet.Common.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            lock (target.Statstring)
            {
                var newStatstring = new byte[target.Statstring.Length];

                using var m = new MemoryStream(newStatstring);
                using var w = new BinaryWriter(m);

                w.Write((UInt32)targetGame);
                w.Write(target.Statstring[4..]);

                if (target.ActiveChannel != null)
                {
                    target.ActiveChannel.UpdateUser(target, newStatstring); // also sets target.Statstring = newStatstring
                }
                else
                {
                    target.Statstring = newStatstring;
                }
            }

            var targetEnv = new Dictionary<string, string>()
            {
                { "accountName", target.Username },
                { "channel", target.ActiveChannel == null ? "(null)" : target.ActiveChannel.Name },
                { "game", Product.ProductName(targetGame, true) },
                { "host", "BNETDocs" },
                { "localTime", target.LocalTime.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "name", target.OnlineName },
                { "onlineName", target.OnlineName },
                { "realm", "BNETDocs" },
                { "realmTime", DateTime.Now.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "realmTimezone", $"UTC{DateTime.Now:zzz}" },
                { "user", target.OnlineName },
                { "username", target.OnlineName },
                { "userName", target.OnlineName },
            };
            var env = targetEnv.Concat(context.Environment);

            r = Resources.AdminSpoofUserGameCommand;

            foreach (var kv in env)
            {
                r = r.Replace("{" + kv.Key + "}", kv.Value);
            }

            foreach (var line in r.Split(Battlenet.Common.NewLine))
            {
                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
            }
        }
    }
}
